// @lynx-js/react/runtime/lib/gesture/processGesture.js
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { onWorkletCtxUpdate } from '@lynx-js/react/worklet-runtime/bindings';
import { GestureTypeInner } from './types.js';
function isSerializedGesture(gesture) {
    return gesture.__isSerialized ?? false;
}
function getGestureInfo(gesture, oldGesture, isFirstScreen, dom) {
    const config = {
        callbacks: [],
    };
    const baseGesture = gesture;
    if (baseGesture.config) {
        config.config = baseGesture.config;
    }
    for (const key of Object.keys(baseGesture.callbacks)) {
        const callback = baseGesture.callbacks[key];
        const oldCallback = oldGesture?.callbacks[key];
        onWorkletCtxUpdate(callback, oldCallback, isFirstScreen, dom);
        config.callbacks.push({
            name: key,
            callback: callback,
        });
    }
    const relationMap = {
        waitFor: baseGesture?.waitFor?.map(subGesture => subGesture.id) ?? [],
        simultaneous: baseGesture?.simultaneousWith?.map(subGesture => subGesture.id) ?? [],
        continueWith: baseGesture?.continueWith?.map(subGesture => subGesture.id) ?? [],
    };
    return {
        config,
        relationMap,
    };
}
export function processGesture(dom, gesture, oldGesture, isFirstScreen, gestureOptions) {
    if (!gesture || !isSerializedGesture(gesture)) {
        return;
    }
    if (!(gestureOptions && gestureOptions.domSet)) {
        __SetAttribute(dom, 'has-react-gesture', true);
        __SetAttribute(dom, 'flatten', false);
    }
    if (gesture.type === GestureTypeInner.COMPOSED) {
        for (const [index, subGesture] of gesture.gestures.entries()) {
            processGesture(dom, subGesture, oldGesture?.gestures[index], isFirstScreen, {
                domSet: true,
            });
        }
    }
    else {
        const baseGesture = gesture;
        const oldBaseGesture = oldGesture;
        const { config, relationMap } = getGestureInfo(baseGesture, oldBaseGesture, isFirstScreen, dom);
        __SetGestureDetector(dom, baseGesture.id, baseGesture.type, config, relationMap);
    }
}
