// @rsdoctor/graph/dist/esm/graph/chunk-graph/graph.mjs
class ChunkGraph {
    getAssets() {
        return Array.from(this._assetMap.values());
    }
    getChunks() {
        return Array.from(this._chunkMap.values());
    }
    addAsset(...assets) {
        assets.forEach((asset)=>{
            this._assetMap.set(asset.id, asset);
            this._assetPathMap.set(asset.path, asset);
        });
    }
    addChunk(...chunks) {
        chunks.forEach((chunk)=>{
            this._chunkMap.set(chunk.id, chunk);
        });
    }
    getChunkById(id) {
        return this._chunkMap.get(id);
    }
    getAssetByPath(path) {
        return this._assetPathMap.get(path);
    }
    getAssetById(id) {
        return this._assetMap.get(id);
    }
    getEntryPoints() {
        return Array.from(this._entrypointMap.values());
    }
    getEntryPointById(id) {
        return this._entrypointMap.get(id);
    }
    addEntryPoint(...entrypoints) {
        entrypoints.forEach((entrypoint)=>{
            this._entrypointMap.set(entrypoint.id, entrypoint);
        });
    }
    toData(type) {
        return {
            assets: Array.from(this._assetMap.values()).map((item)=>item.toData(type)),
            chunks: Array.from(this._chunkMap.values()).map((item)=>item.toData()),
            entrypoints: Array.from(this._entrypointMap.values()).map((item)=>item.toData())
        };
    }
    setChunks(chunks) {
        this._chunkMap = new Map(chunks.map((item)=>[
                item.id,
                item
            ]));
    }
    setEntrypoints(entrypoints) {
        this._entrypointMap = new Map(entrypoints.map((item)=>[
                item.id,
                item
            ]));
    }
    setAssets(assets) {
        this._assetMap = new Map(assets.map((item)=>[
                item.id,
                item
            ]));
        this._assetPathMap = new Map(assets.map((item)=>[
                item.path,
                item
            ]));
    }
    constructor(){
        this._assetMap = new Map();
        this._assetPathMap = new Map();
        this._chunkMap = new Map();
        this._entrypointMap = new Map();
    }
}
export { ChunkGraph };
