// browserslist-to-es-version/dist/index.js
import browserslist from "browserslist";
const ES_VERSIONS_MAP = {
    chrome: [
        51,
        52,
        57,
        64,
        73,
        80,
        85,
        94
    ],
    edge: [
        15,
        15,
        15,
        79,
        79,
        80,
        85,
        94
    ],
    safari: [
        10,
        10.3,
        11,
        16.4,
        17,
        17,
        17,
        17
    ],
    firefox: [
        54,
        54,
        54,
        78,
        78,
        80,
        80,
        93
    ],
    opera: [
        38,
        39,
        44,
        51,
        60,
        67,
        71,
        80
    ],
    samsung: [
        5,
        6.2,
        6.2,
        8.2,
        11.1,
        13,
        14,
        17
    ]
};
const aliases = {
    ios_saf: 'safari',
    and_chr: 'chrome',
    and_ff: 'firefox'
};
const renameBrowser = (name)=>aliases[name] || name;
function browserslistToESVersion(browsers) {
    const projectBrowsers = browserslist(browsers, {
        ignoreUnknownVersions: true
    });
    let esVersion = 2022;
    for (const item of projectBrowsers){
        const pairs = item.split(' ');
        if (pairs.length < 2) continue;
        const browser = renameBrowser(pairs[0]);
        const version = Number(pairs[1].split('-')[0]);
        if (Number.isNaN(version)) continue;
        if ('ie' === browser || 'android' === browser && version < 6) {
            esVersion = 5;
            break;
        }
        const versions = ES_VERSIONS_MAP[browser];
        if (versions) {
            if (version < versions[0]) esVersion = Math.min(5, esVersion);
            else if (version < versions[1]) esVersion = Math.min(2015, esVersion);
            else if (version < versions[2]) esVersion = Math.min(2016, esVersion);
            else if (version < versions[3]) esVersion = Math.min(2017, esVersion);
            else if (version < versions[4]) esVersion = Math.min(2018, esVersion);
            else if (version < versions[5]) esVersion = Math.min(2019, esVersion);
            else if (version < versions[6]) esVersion = Math.min(2020, esVersion);
            else if (version < versions[7]) esVersion = Math.min(2021, esVersion);
        }
    }
    return esVersion;
}
export { browserslistToESVersion };
