// @rsdoctor/utils/dist/esm/rule-utils/parser/parser.mjs
import { Parser } from "acorn";
import { importAttributes } from "acorn-import-attributes";
import { asserts } from "./asserts.mjs";
import * as __WEBPACK_EXTERNAL_MODULE__utils_mjs_25ece7d1__ from "./utils.mjs";
import * as __WEBPACK_EXTERNAL_MODULE_acorn_walk_9c3ac672__ from "acorn-walk";
const acornParserInternal = Parser.extend(importAttributes);
let acornParserExport = Parser.extend(importAttributes);
const parser = {
    walk: __WEBPACK_EXTERNAL_MODULE_acorn_walk_9c3ac672__,
    parse: (input, options)=>acornParserExport.parse(input, options),
    parseExpressionAt: (input, pos, options)=>acornParserExport.parseExpressionAt(input, pos, options),
    extend (...args) {
        acornParserExport = acornParserExport.extend(...args);
        return acornParserExport;
    },
    asserts: asserts,
    utils: __WEBPACK_EXTERNAL_MODULE__utils_mjs_25ece7d1__,
    internal: {
        parse: (input, options)=>acornParserInternal.parse(input, options),
        parseExpressionAt: (input, pos, options)=>acornParserInternal.parseExpressionAt(input, pos, options)
    }
};
export { parser };
