// @lynx-js/preact-devtools/lib/constants.js
export const ClientToDevtools = "preact-client-to-devtools";
export const DevtoolsToClient = "preact-devtools-to-client";
export const ContentScriptName = "preact-content-script";
export const DevtoolsPanelName = "preact-devtools-panel";
export const DevtoolsPanelInlineName = "preact-devtools-panel/inline";
export const PageHookName = "preact-page-hook";
export const PROFILE_RELOAD = "preact-devtools_profile-and-reload";
export const STATS_RELOAD = "preact-devtools_stats-and-reload";
export var NodeType;
(function (NodeType) {
    NodeType[NodeType["Element"] = 1] = "Element";
    NodeType[NodeType["Text"] = 3] = "Text";
    NodeType[NodeType["CData"] = 4] = "CData";
    NodeType[NodeType["XMLProcessingInstruction"] = 7] = "XMLProcessingInstruction";
    NodeType[NodeType["Comment"] = 8] = "Comment";
    NodeType[NodeType["Document"] = 9] = "Document";
    NodeType[NodeType["DocumentType"] = 10] = "DocumentType";
    NodeType[NodeType["DocumentFragment"] = 11] = "DocumentFragment";
})(NodeType || (NodeType = {}));
