// @lynx-js/preact-devtools/lib/view/store/types.js
export var DevNodeType;
(function (DevNodeType) {
    /**
     * Groups are virtual nodes inserted by the devtools
     * to make certain operations easier. They are not
     * created by Preact.
     */
    DevNodeType[DevNodeType["Group"] = 0] = "Group";
    DevNodeType[DevNodeType["Element"] = 1] = "Element";
    DevNodeType[DevNodeType["ClassComponent"] = 2] = "ClassComponent";
    DevNodeType[DevNodeType["FunctionComponent"] = 3] = "FunctionComponent";
    DevNodeType[DevNodeType["ForwardRef"] = 4] = "ForwardRef";
    DevNodeType[DevNodeType["Memo"] = 5] = "Memo";
    DevNodeType[DevNodeType["Suspense"] = 6] = "Suspense";
    DevNodeType[DevNodeType["Context"] = 7] = "Context";
    DevNodeType[DevNodeType["Consumer"] = 8] = "Consumer";
    DevNodeType[DevNodeType["Portal"] = 9] = "Portal";
})(DevNodeType || (DevNodeType = {}));
export var Panel;
(function (Panel) {
    Panel["ELEMENTS"] = "ELEMENTS";
    Panel["PROFILER"] = "PROFILER";
    Panel["SETTINGS"] = "SETTINGS";
    Panel["STATISTICS"] = "STATISTICS";
})(Panel || (Panel = {}));
