// @lynx-js/react/runtime/lib/worklet/workletRef.js
import { WorkletEvents } from '@lynx-js/react/worklet-runtime/bindings';
import { addWorkletRefInitValue } from './workletRefPool.js';
import { useMemo } from '../hooks/react.js';
// Split into two variables for testing purposes
let lastIdBG = 0;
let lastIdMT = 0;
export function clearWorkletRefLastIdForTesting() {
    lastIdBG = lastIdMT = 0;
}
class WorkletRef {
    /**
     * @internal
     */
    _wvid;
    /**
     * @internal
     */
    _initValue;
    /**
     * @internal
     */
    _type;
    /**
     * @internal
     */
    _lifecycleObserver;
    /**
     * @internal
     */
    constructor(initValue, type) {
        this._initValue = initValue;
        this._type = type;
        if (__JS__) {
            this._wvid = ++lastIdBG;
            addWorkletRefInitValue(this._wvid, initValue);
        }
        else {
            this._wvid = --lastIdMT;
        }
    }
    get current() {
        if (__JS__ && __DEV__) {
            throw new Error('MainThreadRef: value of a MainThreadRef cannot be accessed in the background thread.');
        }
        if (__LEPUS__ && __DEV__) {
            /* v8 ignore next 3 */
            throw new Error('MainThreadRef: value of a MainThreadRef cannot be accessed outside of main thread script.');
        }
        return undefined;
    }
    set current(_) {
        if (__JS__ && __DEV__) {
            throw new Error('MainThreadRef: value of a MainThreadRef cannot be accessed in the background thread.');
        }
        if (__LEPUS__ && __DEV__) {
            throw new Error('MainThreadRef: value of a MainThreadRef cannot be accessed outside of main thread script.');
        }
    }
    /**
     * @internal
     */
    toJSON() {
        return {
            _wvid: this._wvid,
        };
    }
}
/**
 * A `MainThreadRef` is a ref that can only be accessed on the main thread. It is used to preserve
 * states between main thread function calls.
 * The data saved in `current` property of the `MainThreadRef` can be read and written in
 * multiple main thread functions.
 * @public
 */
export class MainThreadRef extends WorkletRef {
    constructor(initValue) {
        super(initValue, 'main-thread');
        if (__JS__) {
            const id = this._wvid;
            this._lifecycleObserver = lynx.getNativeApp().createJSObjectDestructionObserver?.(() => {
                lynx.getCoreContext?.().dispatchEvent({
                    type: WorkletEvents.releaseWorkletRef,
                    data: {
                        id,
                    },
                });
            });
        }
    }
}
export function useMainThreadRef(initValue) {
    return useMemo(() => {
        return new MainThreadRef(initValue);
    }, []);
}
