// @rsdoctor/sdk/dist/esm/sdk/server/apis/graph.mjs
import { SDK } from "@rsdoctor/types";
import { BaseAPI } from "./base.mjs";
import { Router } from "../router.mjs";
function _ts_decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : null === desc ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if ("object" == typeof Reflect && "function" == typeof Reflect.decorate) r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function _ts_metadata(k, v) {
    if ("object" == typeof Reflect && "function" == typeof Reflect.metadata) return Reflect.metadata(k, v);
}
class GraphAPI extends BaseAPI {
    async getAssetsSummary() {
        const { req } = this.ctx;
        const { withFileContent } = req.body;
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetAssetsSummary, {
            withFileContent
        });
    }
    async getModuleGraphForAsset() {
        const { req } = this.ctx;
        const { assetPath } = req.body;
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetAssetDetails, {
            assetPath
        });
    }
    async getChunksByModuleId() {
        const { req } = this.ctx;
        const { moduleId } = req.body;
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetChunksByModuleId, {
            moduleId
        });
    }
    async getModuleDetails() {
        const { req } = this.ctx;
        const { moduleId } = req.body;
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetModuleDetails, {
            moduleId
        });
    }
    async getDependencyByResolvedRequest() {
        const { req } = this.ctx;
        const { moduleIds = [] } = req.body;
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetModulesByModuleIds, {
            moduleIds
        });
    }
    async getModuleCodeByModuleId() {
        const { req } = this.ctx;
        const { moduleId } = req.body;
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetModuleCodeByModuleId, {
            moduleId
        });
    }
    async getModuleCodeByModuleIds() {
        const { req } = this.ctx;
        const { moduleIds } = req.body;
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetModuleCodeByModuleIds, {
            moduleIds
        });
    }
    async getEntryModulesSummary() {
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetEntryPoints);
    }
}
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetAssetsSummary),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getAssetsSummary", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetAssetDetails),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getModuleGraphForAsset", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetChunksByModuleId),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getChunksByModuleId", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetModuleDetails),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getModuleDetails", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetModulesByModuleIds),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getDependencyByResolvedRequest", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetModuleCodeByModuleId),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getModuleCodeByModuleId", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetModuleCodeByModuleIds),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getModuleCodeByModuleIds", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetEntryPoints),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], GraphAPI.prototype, "getEntryModulesSummary", null);
export { GraphAPI };
