// @rsdoctor/sdk/dist/esm/sdk/server/apis/project.mjs
import { SDK } from "@rsdoctor/types";
import { BaseAPI } from "./base.mjs";
import { Router } from "../router.mjs";
import { getLocalIpAddress } from "../utils.mjs";
function _ts_decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : null === desc ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if ("object" == typeof Reflect && "function" == typeof Reflect.decorate) r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}
function _ts_metadata(k, v) {
    if ("object" == typeof Reflect && "function" == typeof Reflect.metadata) return Reflect.metadata(k, v);
}
class ProjectAPI extends BaseAPI {
    async env() {
        const { server } = this.ctx;
        return {
            ip: getLocalIpAddress(),
            port: server.port
        };
    }
    async manifest() {
        const { sdk } = this.ctx;
        const data = sdk.getManifestData();
        return JSON.stringify(data);
    }
    async getProjectInfo() {
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetProjectInfo);
    }
    async getClientRoutes() {
        return this.dataLoader.loadAPI(SDK.ServerAPI.API.GetClientRoutes);
    }
}
_ts_decorate([
    Router.get(SDK.ServerAPI.API.Env),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], ProjectAPI.prototype, "env", null);
_ts_decorate([
    Router.get(SDK.ServerAPI.API.Manifest),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], ProjectAPI.prototype, "manifest", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetProjectInfo),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], ProjectAPI.prototype, "getProjectInfo", null);
_ts_decorate([
    Router.post(SDK.ServerAPI.API.GetClientRoutes),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", []),
    _ts_metadata("design:returntype", Promise)
], ProjectAPI.prototype, "getClientRoutes", null);
export { ProjectAPI };
